### PR TITLE
fix(ts): Typescript 4 fixes for HoF

### PR DESCRIPTION
**Pull request by scttcper**
edit:

Here's a [typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4BYAKCrQBsUBnBuADTbiQA8YkA7AE2bJ0MAHQBhXJF58YAHgDedRgwByKEEgBccBjCjBeAcwC+AGjhL6TdZp16DxkwD5LVOB90wUPOAF5LZRsNbQIUACM0fBMKakpPOCIBJCgACgBKN3iEzyIYAFcoXjg5fmAANzggtRC-BRgAC2AGUT0fJFFq2yQTS1F+xubRMBwwBhc5AHoy8ud3TxMqRcogA) of the issue.

Tried to declare the props similar to how `object.assign` was functioning.
I don't like that typescript is enforcing this. This feels like a lint error and not a type error.

### CODE REVIEW

**Review by billyvg — COMMENTED** [review]
Are they being declared twice? If so can we Exclude one of them?

Does the optional affect the wrapped components?

File: src/sentry/static/sentry/app/utils/withOrganizations.tsx [hunk 1]
```diff
@@ -8,7 +8,7 @@ import {OrganizationSummary} from 'app/types';
 
 type InjectedOrganizationsProps = {
   organizationsLoading?: boolean;
-  organizations: OrganizationSummary[];
+  organizations?: OrganizationSummary[];
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/utils/withOrganizations.tsx:11]: in this case you could do like below:

```
type InjectedOrganizationsProps = {
  organizationsLoading?: boolean;
  organizations: OrganizationSummary[];
};

type State = Pick<InjectedOrganizationsProps, 'organizations'>;
```

**Comment by scttcper** [comment]
Here's a [typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4BYAKCrQBsUBnBuADTbiQA8YkA7AE2bJ0MAHQBhXJF58YAHgDedRgwByKEEgBccBjCjBeAcwC+AGjhL6TdZp16DxkwD5LVOB90wUPOAF5LZRsNbQIUACM0fBMKakpPOCIBJCgACgBKN3iEzyIYAFcoXjg5fmAANzggtRC-BRgAC2AGUT0fJFFq2yQTS1F+xubRMBwwBhc5AHoy8ud3TxMqRcogA) of the issue.

I think changing the props was wrong. I've now tried to declare the props similar to how object.assign would have

**Review by dashed — APPROVED** [review]

**Comment by priscilawebdev** [comment]
@scttcper the problem in the example is that props and state have the same property "className" and when you use ...this.props after className={this.state.className} it will overwrite className. I played a bit and if you inverse the order, the bug is gone https://www.typescriptlang.org/play?ts=4.0.2#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4BYAKCrQBsUBnBuADTbiQA8YkA7AE2bJ0MAHQBhXJF58YAHgDedRgwByKEEgBccBjCjBeAcwC+AGjhL6TdZp16DxkwD5LVOB90wUPOAF5LZRsNbQIUACM0fBMKakpPOCIBJCgACgBKN3iEzyIYAFcoXjg5fmAAN0tRapgAC2AGUTAcMAYTOCC1EL8FOobRPR8kUU7bJHbnOQB6MvLnd08TKhMgA

File: src/sentry/static/sentry/app/utils/withApi.tsx [hunk 2]
```diff
@@ -46,7 +46,8 @@ const withApi = <P extends InjectedApiProps>(
     private api: Client;
 
     render() {
-      return <WrappedComponent api={this.api} {...(this.props as P)} />;
+      const {api = this.api} = this.props as P;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/utils/withApi.tsx:49]: here you are destructing props

File: src/sentry/static/sentry/app/utils/withApi.tsx [hunk 3]
```diff
@@ -46,7 +46,8 @@ const withApi = <P extends InjectedApiProps>(
...
+      return <WrappedComponent {...(this.props as P)} api={api} />;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/utils/withApi.tsx:50]: but here you are passing everything twice again.. it's fine as the `api` prop comes after `this.props`, but I I would rather write like:

 ```
 const {api, ...props} = this.props;
 return <WrappedComponent {...({api: api ?? this.api, ...props} as P)} />;
```

Comment (scttcper) [reply src/sentry/static/sentry/app/utils/withApi.tsx]: changed it over to this style

**Comment by dashed** [comment]
I'm curious, is this the same typedef pattern that other higher-order components use?

File: src/sentry/static/sentry/app/utils/withSavedSearches.tsx [hunk 4]
```diff
@@ -10,7 +11,7 @@ type InjectedSavedSearchesProps = {
   savedSearches: SavedSearch[];
   savedSearchLoading: boolean;
   savedSearch: SavedSearch | null;
-};
+} & RouteComponentProps<{searchId?: string}, {}>;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/utils/withSavedSearches.tsx:14]: 👏 good catch

**Comment by priscilawebdev** [comment]
Everything looks good now 👏🚀

**Pull request merged**